cpp/dolfinx/io/ADIOS2Writers.h

```diff
@@ -867,6 +867,333 @@ void vtx_write_mesh_from_space(adios2::IO& io, adios2::Engine& engine,
+  std::vector<T> geometry_b(num_xdofs_local * gdim);
+  stdex::mdspan<T, dextent> geometry_md(geometry_b.data(), num_xdofs_local,
+                                        gdim);
+  for (std::size_t i = 0; i < num_xdofs_local; ++i)
+    for (std::size_t j = 0; j < gdim; ++j)
+      geometry_md(i, j) = x[3 * i + j];
+
+  // Define geometry variable and "put"
+  adios2::Variable<T> local_geometry = impl_adios2::define_variable<T>(
+      io, "geometry", {num_xdofs_global, gdim}, {local_range[0], 0},
+      {num_xdofs_local, gdim});
+  engine.Put<T>(local_geometry, geometry_b.data());
+
+  // Put topology (connectivity)
+  auto topology = mesh.topology_mutable();
+  assert(topology);
+
+  // Extract cell information
+  auto cell_imap = topology->index_map(topology->dim());
+  std::int32_t num_cells_local = cell_imap->size_local();
+  std::int64_t num_cells_global = cell_imap->size_global();
+  std::array<std::int64_t, 2> local_range_cell = cell_imap->local_range();
+
+  // Extract geometry information
+  if (geometry.cmaps().size() > 1)
+    throw std::runtime_error("Multiple coordinate maps not supported");
+  fem::ElementDofLayout geom_layout = geometry.cmaps()[0].create_dof_layout();
+  int num_dofs_per_cell = geom_layout.num_entity_closure_dofs(geometry.dim());
+  auto geometry_imap = geometry.index_map();
+  std::experimental::mdspan<const std::int32_t,
+                            std::experimental::dextents<std::size_t, 2>>
+      geometry_dmap = geometry.dofmap();
+  // Convert local geometry to global geometry
+  std::vector<std::int64_t> connectivity_out(num_cells_local
+                                             * num_dofs_per_cell);
+
+  std::span<const std::int32_t> local_dmap(geometry_dmap.data_handle(),
+                                           num_cells_local * num_dofs_per_cell);
+  geometry_imap->local_to_global(local_dmap, connectivity_out);
+
+  // Put local topology in global dataset
+  adios2::Variable<std::int64_t> local_topology
+      = impl_adios2::define_variable<std::int64_t>(
+          io, "connectivity",
+          {(std::size_t)num_cells_global, (std::size_t)num_dofs_per_cell},
+          {(std::size_t)local_range_cell[0], (std::size_t)0},
+          {(std::size_t)num_cells_local, (std::size_t)num_dofs_per_cell});
+  engine.Put<std::int64_t>(local_topology, connectivity_out.data());
+
+  // Compute mesh permutations
+  topology->create_entity_permutations();
+  const std::vector<std::uint32_t>& cell_perm
+      = topology->get_cell_permutation_info();
+  adios2::Variable<std::uint32_t> local_cell_perm
+      = impl_adios2::define_variable<std::uint32_t>(
+          io, "CellPermutations", {(std::size_t)num_cells_global},
+          {(std::size_t)local_range_cell[0]}, {(std::size_t)num_cells_local});
+  engine.Put<std::uint32_t>(local_cell_perm, cell_perm.data());
+
+  engine.PerformPuts();
+}
+
+} // namespace impl_checkpoint
+
+template <std::floating_point T>
+class ADIOS2Checkpointer
```

Comment:
In all likelihood, we'll not use anything other than ADIOS2 for checkpointing, so probably don't need ADIOS2 in the name.

Also, code only seems to handle meshes so name is a bit misleading.